### PR TITLE
Update loop unwinding order

### PR DESCRIPTION
**Pull request by feliperodri**
# Change overview
We update the Makefile.common so that :
- When function contracts with DFCC is used, the CPROVER library must be linked and all loops that need unwinding be unwound  before contract instrumentation
- When the old function contracts or loop contracts are used the library must be linked after contract instrumentation. Loops in user code must be unwound before contract instrumentation and loops in CPROVER library functions unwound after contract instrumentation
- When no contracts are used, the CPROVER library and loop unwinding are performed a the last step of the build when cbmc is invoked.

# CPROVER Library flags management

The flags configuring the CPROVER library (i.e., CBMC_OPT_CONFIG_LIBRARY) must be passed at any step where the library is be loaded.

- When linking with goto instrument `goto-instrument --add-library $(CBMC_OPT_CONFIG_LIBRARY) ...`
- When applying contracts with DFCC `goto-instrument  $(CBMC_OPT_CONFIG_LIBRARY) --dfcc harness...` because the behavior  of the built-in `__CPROVER_is_fresh` is also configured by malloc failure modes
- When linking to the library at analysis time `cbmc $(CBMC_OPT_CONFIG_LIBRARY) ...`

# Loop unwinding

Loop unwinding is parameterized by the following  envs:
- `UNWINDSET` : loops to unwind in user code (i.e., proof and project code);
- `CPROVER_LIBRARY_UNWINDSET` : loops to unwind in CPROVER library functions; 
- `DEFAULT_UNWIND`: `--unwind 1` (sets a default unwinding depth for all loops of a model)
- `STRING_ABSTRACTION`: --string-abstraction (abstracts away loops in `string.h` functions when set)

Loops are handled as follows:
- loops in user code: 
  - The subset of user loops to manually unwind is specified in `UNWINDSET`;
  - Loops that have loop contracts and are not listed in `UNWINDSET` will be abstracted (i.e., `--apply-loop-contracts`);
  - The rest of the loops is handled using `DEFAULT_UNWIND`
- loops in CPROVER library.
  - The subset of CPROVER library loops to unwind is specified in `CPROVER_LIBRARY_UNWINDSET`;
  - The rest of the loops is handled using `DEFAULT_UNWIND` or by using `STRING_ABSTRACTION` (e.g., `strlen`);
- loops in contract instrumentation functions on DFCC mode. These loops *must* be dynamically unwound during analysis;

## When no contracts are used

1. Early unwind. We call `goto-instrument $(UNWINDSET) $(CBMC_FLAG_UNWINDING_ASSERTIONS)` to unwind specific subsets of loops in user code.

2. Linking CPROVER library and loop unwinding for all remaining loops happens at the same time when invoking CBMC (i.e., property checking, coverage checking and printing safety properties).

```Makefile
cbmc $(CBMC_VERBOSITY) $(CBMCFLAGS) $(CHECKFLAGS)
      $(CBMC_OPT_CONFIG_LIBRARY)
      $(DEFAULT_UNWIND) $(CPROVER_LIBRARY_UNWINDSET) $(CBMC_FLAG_UNWINDING_ASSERTIONS)
      --trace --xml-ui
```

## When old function contracts or loop contracts are used

After restricting function pointers:
1. Early unwind. We call `goto-instrument $(UNWINDSET) $(CBMC_FLAG_UNWINDING_ASSERTIONS)` to unwind specific subsets of loops in user code.
3. Contracts instrumentation. We call `goto-instrument --apply-loop-contracts --enforce-contract ...` to instrument the model. This abstracts loops that have contracts, as well as loops that are in functions that get replaced by contracts.
4. Clean up global variables. We call `goto-instrument --slice-global-init ...`
5. Clean up unused functions. We call `goto-instrument --drop-unused-functions ...`
8. Invoke CBMC.

```Makefile
cbmc $(CBMC_VERBOSITY) $(CBMCFLAGS) $(CHECKFLAGS)
      $(CBMC_OPT_CONFIG_LIBRARY)
      $(DEFAULT_UNWIND) $(CPROVER_LIBRARY_UNWINDSET) $(CBMC_FLAG_UNWINDING_ASSERTIONS)
      --trace --xml-ui
```

## When function contracts with DFCC are used

After restricting function pointers:
1. Link CPROVER library. We call `goto-instrument --add-library  $(CBMC_OPT_CONFIG_LIBRARY) ...`. to link the model to the CPROVER standard library;
6. Early unwind. We call `goto-instrument $(CPROVER_LIBRARY_UNWINDSET) $(UNWINDSET) $(DEFAULT_UNWIND) $(CBMC_FLAG_UNWINDING_ASSERTIONS)`;
7. Contracts instrumentation with DFCC. We call `goto-instrument $(CBMC_OPT_CONFIG_LIBRARY) --dfcc harness ...`. The loops found in `__CPROVER_contracts`  *must* be dynamically unwound during analysis which rules out using `--unwind 1` with CBMC. DFCC mode also havocs the global state and drops unused functions;
8. Invoke CBMC.

```Makefile
cbmc $(CBMC_VERBOSITY) $(CBMCFLAGS) $(CHECKFLAGS)
      $(CBMC_OPT_CONFIG_LIBRARY)
      $(CBMC_FLAG_UNWINDING_ASSERTIONS)
      --trace --xml-ui
```

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by remi-delmas-3000 — APPROVED** [review]

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common [hunk 1]
```diff
@@ -682,75 +685,76 @@ else
 	  --description "$(PROOF_UID): setting static variables to nondet"
 endif
 
-# Omit unused functions (sharpens coverage calculations)
+# Omit initialization of unused global variables (reduces problem size)
```
Comment (remi-delmas-3000) [inline src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:688]: we should move this after contract instrumentation, same step as "omit unused functions"

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common [hunk 2]
```diff
@@ -682,75 +685,76 @@ else
...
-	  --stdout-file $(LOGDIR)/drop_unused_functions-log.txt \
+	  --stdout-file $(LOGDIR)/slice_global_inits-log.txt \
 	  --pipeline-name "$(PROOF_UID)" \
 	  --ci-stage build \
-	  --description "$(PROOF_UID): dropping unused functions"
+	  --description "$(PROOF_UID): slicing global initializations"
 endif
 
-# Omit initialization of unused global variables (reduces problem size)
+# Early unwind loops (e.g., for C library functions)
```
Comment (remi-delmas-3000) [inline src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:711]: Suggested comment:  "link CPROVER library and unwind all loops (DFCC on) or unwind loops in user code only (DFCC off)"

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common [hunk 3]
```diff
@@ -765,6 +769,10 @@ $(HARNESS_GOTO).goto: $(HARNESS_GOTO)7.goto
 ################################################################
 # Targets to run the analysis commands
 
+ifeq ($(strip $(DFCC_MODE)),)
```
Comment (remi-delmas-3000) [inline src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:772]: Yes !

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common [hunk 4]
```diff
@@ -682,75 +685,76 @@ else
...
+	  --description "$(PROOF_UID): unwinding C library loops during instrumentation"
 endif
 
-# Early unwind loops (e.g., for functions annotated with contracts)
+# Replace function contracts, check function contracts, instrument for loop contracts
 $(HARNESS_GOTO)6.goto: $(HARNESS_GOTO)5.goto
 	$(LITANI) add-job \
 	  --command \
-	    '$(GOTO_INSTRUMENT) $(CBMC_VERBOSITY) $(CBMC_EARLY_UNWINDSET) $(CBMC_FLAG_UNWINDING_ASSERTIONS) $^ $@' \
+	    '$(GOTO_INSTRUMENT) $(CBMC_DFCC_MODE) $(CBMC_VERBOSITY) $(CBMC_CHECK_FUNCTION_CONTRACTS) $(CBMC_CHECK_FUNCTION_CONTRACTS_REC) $(CBMC_USE_FUNCTION_CONTRACTS) $(CBMC_APPLY_LOOP_CONTRACTS) $^ $@' \
```
Comment (remi-delmas-3000) [inline src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:739]: we are missing the `--nondet-static-exclude VAR` flags here, cf.`$(NONDET_STATIC)`

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common [hunk 5]
```diff
@@ -682,75 +685,76 @@ else
...
+	  --stdout-file $(LOGDIR)/unwind_loops-log.txt \
 	  --pipeline-name "$(PROOF_UID)" \
 	  --ci-stage build \
-	  --description "$(PROOF_UID): not slicing global initializations"
+	  --description "$(PROOF_UID): unwinding all loops during instrumentation"
 else
 	$(LITANI) add-job \
 	  --command \
-	    '$(GOTO_INSTRUMENT) $(CBMC_VERBOSITY) --slice-global-inits $^ $@' \
+	    '$(GOTO_INSTRUMENT) $(CBMC_VERBOSITY) $(ADD_LIBRARY_FLAG) $(CBMC_EARLY_UNWINDSET) $(CBMC_FLAG_UNWINDING_ASSERTIONS) $^ $@' \
```
Comment (remi-delmas-3000) [inline src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:726]: When DFCC is not used, we must not add the CPROVER library at this stage, instead it needs to be added after instrumentation has happened, when calling `cbmc`

**Review by remi-delmas-3000 — CHANGES_REQUESTED** [review]
There is an off-by one difference in interpretation of the unwinding bound value between cbmc and goto-instrument. Since the UNWINDSET is now passed to either of them depending on wether DFCC is used, do you think we should automatically adjust the value so it means the same thing in both cases ? 

do we also need to propagate the changes to  the following target ?
```
$(LOGDIR)/coverage.xml: $(HARNESS_GOTO).goto
```

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common (hunk 1 above)
Comment (feliperodri) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:688]: Implemented as suggested.

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common (hunk 2 above)
Comment (feliperodri) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common]: I've updated the comment.

**Comment by feliperodri** [comment]
> There is an off-by one difference in interpretation of the unwinding bound value between cbmc and goto-instrument. Since the UNWINDSET is now passed to either of them depending on wether DFCC is used, do you think we should automatically adjust the value so it means the same thing in both cases ?
> 
> do we also need to propagate the changes to the following target ?
> 
> ```
> $(LOGDIR)/coverage.xml: $(HARNESS_GOTO).goto
> ```

So I wrote this little program.
```c
int main()
{
  int i = 0;
  int j = 10;
  while(i < j)
  {
    i++;
  }
}
```
If we unroll the loop and verify it with CBMC as follow, we get an assertion violation.
```
cbmc main.c --unwinding-assertions --unwind 10
```
If we unroll the loop using `goto-instrument` using the same bound (as follow), the verification succeeds.
```
goto-cc main.c -o main; goto-instrument main main-inst --unwind 10 --unwinding-assertions; cbmc main-inst
```

**Takeaway** Users now add +1 to their bound, if our change move unrolling to `goto-instrument`, there should be no unwinding failures. I’d document this off by 1 difference, but not adjust the value.

**Comment by feliperodri** [comment]
@remi-delmas-3000 I tested the new `Makefile.common` with the following proofs:
1. Proofs from `coreHTTP` and `coreJSON` without any contracts;
2. Proofs from `coreJSON` with contracts on DFCC mode;
3. Proofs from `coreJSON` off DFCC mode;
4. Proofs from AWS C Common with loop contracts;

**Review by remi-delmas-3000 — APPROVED** [review]

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common [hunk 6]
```diff
@@ -341,9 +344,9 @@ UNWINDSET ?=
 # contracts).  To satisfy this requirement, it may be necessary to
 # unwind some loops before the function contract and loop invariant
 # transformations are applied to the goto program.  This variable
-# EARLY_UNWINDSET is identical to UNWINDSET, and we assume that the
-# loops mentioned in EARLY_UNWINDSET and UNWINDSET are disjoint.
-EARLY_UNWINDSET ?=
+# CPROVER_LIBRARY_UNWINDSET is identical to UNWINDSET, and we assume that the
+# loops mentioned in CPROVER_LIBRARY_UNWINDSET and UNWINDSET are disjoint.
+CPROVER_LIBRARY_UNWINDSET ?=
```
Comment (karkhaz) [inline src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:349]: 🍓 This changes a variable that users can set in their Makefiles and thus is likely to break somebody's proof. Why is this change needed?

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common [hunk 7]
```diff
@@ -393,10 +396,15 @@ CBMC_TIMEOUT ?= 21600
 # Replace all uses of char * by a struct that carries that string,
 # and also the underlying allocation and the C string length.
 STRING_ABSTRACTION ?=
+ifdef STRING_ABSTRACTION
```
Comment (karkhaz) [inline src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:399]: 🥝 Why do we need both `ifdef` and `ifneq`? I think `ifneq` is sufficient and consistent with what we do in the rest of the file

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common [hunk 8]
```diff
@@ -252,12 +252,12 @@ CBMC_FLAG_SIGNED_OVERFLOW_CHECK ?= --signed-overflow-check
 CBMC_FLAG_UNDEFINED_SHIFT_CHECK ?= --undefined-shift-check
 CBMC_FLAG_UNSIGNED_OVERFLOW_CHECK ?= --unsigned-overflow-check
 CBMC_FLAG_UNWINDING_ASSERTIONS ?= --unwinding-assertions
-CBMC_FLAG_UNWIND ?= --unwind 1
+CBMC_DEFAULT_UNWIND ?= --unwind 1
```
Comment (karkhaz) [inline src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:255]: 🍓

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common [hunk 9]
```diff
@@ -303,6 +303,9 @@ COMPILE_FLAGS ?= -Wall
 LINK_FLAGS ?= -Wall
 EXPORT_FILE_LOCAL_SYMBOLS ?= --export-file-local-symbols
 
+# During instrumentation, it adds models of C library functions
+ADD_LIBRARY_FLAG ?= --add-library
```
Comment (karkhaz) [inline src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:307]: If this is supposed to be switched on all the time, please use `:=`.

If users should be able to switch it off, please do
```
  ifeq ($(strip $(NO_ADD_LIBRARY_FLAG)),)
    CBMC_ADD_LIBRARY_FLAG := --add-library
  else
    CBMC_ADD_LIBRARY_FLAG := 
  endif
```

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common [hunk 10]
```diff
@@ -418,13 +426,15 @@ CBMC_CHECK_FUNCTION_CONTRACTS_REC := $(patsubst %,--enforce-contract-rec %, $(CH
...
+
 # Proof writers may also apply function contracts using the Dynamic Frame
 # Condition Checking (DFCC) mode. For more information on DFCC,
 # please see https://diffblue.github.io/cbmc/contracts-dev-spec-dfcc.html.
-DFCC_MODE ?=
-ifdef DFCC_MODE
-  ifneq ($(strip $(DFCC_MODE)),)
-    CBMC_DFCC_MODE := --dfcc $(HARNESS_ENTRY) $(CBMC_OPT_CONFIG_LIBRARY)
+USE_DYNAMIC_FRAMES ?=
+ifdef USE_DYNAMIC_FRAMES
```
Comment (karkhaz) [inline src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:435]: 🥝

**Review by karkhaz — CHANGES_REQUESTED** [review]

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common (hunk 6 above)
Comment (jimgrundy) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:349]: Is this orthogonal to the changes needed for contracts and DFCC? If so, perhaps it could be offered as a separate PR with explanation for the motivation?

Comment (jimgrundy) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:349]: I believe Mark had a similar concern, so if it could be factored out it might progress more smoothly.

Comment (feliperodri) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:349]: Nobody uses this variables in their proofs today. This change is necessary cause we may need to unroll loops for CPROVER library (e.g., models for `strlen` and `memcpy`) at different stages depending whether users will use contracts or not. For instance, using dynamic frames, users need to unroll everything during instrumentation; however, if no contracts are used, then they may unroll these loops when invoking CBMC for property checking.

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common (hunk 8 above)
Comment (feliperodri) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:255]: Same answer as 🍓. Plus, this is an internal variable that it's only used in the `Makefile.common`.

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common (hunk 6 above)
Comment (karkhaz) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:349]: OK

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common (hunk 7 above)
Comment (feliperodri) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:399]: I was following the same pattern used for other variables. I thought the idea was to avoid the comparing and `strip` at all if the variable is not even defined.

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common (hunk 10 above)
Comment (feliperodri) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:435]: Same answer for 🥝.

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common (hunk 7 above)
Comment (feliperodri) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:399]: For example,
```Makefile
ifdef RESTRICT_FUNCTION_POINTER
  ifneq ($(strip $(RESTRICT_FUNCTION_POINTER)),)
    CBMC_RESTRICT_FUNCTION_POINTER := $(patsubst %,--restrict-function-pointer %, $(RESTRICT_FUNCTION_POINTER))
  endif
endif
```

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common (hunk 9 above)
Comment (feliperodri) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common]: Implemented using `:=` as suggested.

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common (hunk 8 above)
Comment (karkhaz) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:255]: It actually was never used in Makefile.common before. So we are now passing `--unwind 1` where we were not before.

I'm very confused.

How did the proofs ever work when --unwind 1 was not being passed to them?

It seems that there is now a risk that we break proofs by newly passing --unwind 1 because the proofs did not have that restriction before, and so adding this unwind bound may cause an unwinding assertion to fail?

**Review by karkhaz — CHANGES_REQUESTED** [review]

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common (hunk 8 above)
Comment (feliperodri) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:255]: @karkhaz It was always used before. This recent [commit](https://github.com/model-checking/cbmc-starter-kit/commit/36acd4385cd01a3ee449b334bdf9ac472146a0b1) that removed from our `Makefile.common`.

Comment (feliperodri) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:255]: To be more specific, this [commit](https://github.com/model-checking/cbmc-starter-kit/pull/150/commits/691274ad64b2b4146f1e3116ea0cf832ae7b046b).

**Review by karkhaz — APPROVED** [review]

**Review by jimgrundy — APPROVED** [review]
I'm content now.

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common (hunk 8 above)
Comment (remi-delmas-3000) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:255]: In earlier versions of the starter kit (up to the one before this PR)

The `--unwind 1` flag was passed to `cbmc` trough in `CBMC_FLAGS`

```
CBMC_FLAG_UNWIND ?= --unwind 1
...
CBMCFLAGS += $(CBMC_FLAG_UNWIND) $(CBMC_UNWINDSET) $(CBMC_FLAG_FLUSH)
```

And the unwinding assertions flag was passed together with CHECK_FLAGS
```
CBMC_FLAG_UNWINDING_ASSERTIONS ?= --unwinding-assertions
...
CHECKFLAGS += $(CBMC_FLAG_UNWINDING_ASSERTIONS)
```

So all proofs were running using `--unwind 1` as default unless they were already processed using `UNWINDSET`.   

In this PR this is still how things happen when no contracts are used. 

When old contracts are used, user loops are unwound before contract instrumentation, and CPROVER library loops are unwound by cbmc at the end, and -- unwind 1 is also given to cbmc.

When DFCC contracts are used, the user loops and CPROVER library loops are unwound before contract instrumentation and --unwind 1 is applied as well before contract instrumentation, which results in all loops being bounded just like when no contracts are used or when old contracts are used, except for the loops introduced by the dynamic frames instrumentation which are dynamically unwound during symex (and always terminate).

**Review by remi-delmas-3000 — APPROVED** [review]

**Review by tautschnig — COMMENTED** [review]
See my comments for specific concerns. Overall, however, I am really concerned about the complexity we have arrived at. In my opinion, we should have completely separate rule sets for the three modes (no contracts, contracts, dfcc) by just putting the rules in three different Makefiles. Depending on the flags set we would then source only one of the three files.

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common [hunk 11]
```diff
@@ -682,78 +685,100 @@ else
...
 
-# Omit unused functions (sharpens coverage calculations)
+# Link CPROVER library if DFCC mode is on
 $(HARNESS_GOTO)4.goto: $(HARNESS_GOTO)3.goto
-ifneq ($(strip $(DFCC_MODE)),)
+ifneq ($(strip $(USE_DYNAMIC_FRAMES)),)
 	$(LITANI) add-job \
-	  --command 'cp $^ $@' \
+	  --command \
+	    '$(GOTO_INSTRUMENT) $(CBMC_VERBOSITY) $(ADD_LIBRARY_FLAG) $(CBMC_OPT_CONFIG_LIBRARY) $^ $@' \
```
[the remark this answers is not in the document]
Comment (remi-delmas-3000) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:693]: yes they are still going to work as intended. When using contracts for function pointers, function pointer restrictions are expected to contain the `function_pointer -> contract_function` mapping explicitly. This will have to be changed later when function pointers get resolved during symex.

**Comment by feliperodri** [comment]
> See my comments for specific concerns. Overall, however, I am really concerned about the complexity we have arrived at. In my opinion, we should have completely separate rule sets for the three modes (no contracts, contracts, dfcc) by just putting the rules in three different Makefiles. Depending on the flags set we would then source only one of the three files.

@tautschnig I opened https://github.com/model-checking/cbmc-starter-kit/issues/160 issue, so we can refactor `Makefile.common` on a separate PR to reduce its complexity.

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common [hunk 12]
```diff
@@ -489,14 +499,18 @@ DEFINES += -DCBMC_OBJECT_BITS=$(CBMC_OBJECT_BITS)
 DEFINES += -DCBMC_MAX_OBJECT_SIZE="(SIZE_MAX>>(CBMC_OBJECT_BITS+1))"
 
 # CI currently assumes cbmc invocation has at most one --unwindset
+
+# UNWINDSET is design for user code (i.e., proof and project code)
```
[the remark this answers is not in the document]
Comment (feliperodri) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common]: Fixed typo.

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common [hunk 13]
```diff
@@ -489,14 +499,18 @@ DEFINES += -DCBMC_OBJECT_BITS=$(CBMC_OBJECT_BITS)
...
 ifdef UNWINDSET
   ifneq ($(strip $(UNWINDSET)),)
     CBMC_UNWINDSET := --unwindset $(subst $(SPACE),$(COMMA),$(strip $(UNWINDSET)))
   endif
 endif
-ifdef EARLY_UNWINDSET
-  ifneq ($(strip $(EARLY_UNWINDSET)),)
-    CBMC_EARLY_UNWINDSET := --unwindset $(subst $(SPACE),$(COMMA),$(strip $(EARLY_UNWINDSET)))
+
+# CPROVER_LIBRARY_UNWINDSET is design for CPROVER library functions
```
[the remark this answers is not in the document]
Comment (feliperodri) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common]: Fixed typo.
